### DISCUSSION TITLE
Timer with ThreadSleepNoEarlyWake

### DISCUSSION THREAD

**Discussion by PhilDay-CT** [CHERIoT-Platform/cheriot-rtos#262] [Q&A] [answered]
As part of a demo I have various threads that I want to sleep for a couple of seconds.  The code is basically something like the following and is running the latest version of the dev container.

```
void __cheri_compartment("timer") init()
{
	int loop = 1;
	while (true)
	{
		Timeout t1{1000};
		thread_sleep(&t1);
		Debug::log("thread {} Woke {}", thread_id_get(), loop++);
	}
}
```
which was giving me a sleep of about 1 second.   

Having recently rebased my repo the timers are now returning more or less immediately.  Turning on debugging in the scheduler I can see that they are waking early:
```
Scheduler: Created thread for trusted stack 0x80000230 (v:1 0x80000230-0x80000338 l:0x108 o:0xa p: G RWcgml -- ---)
Scheduler: Thread 0x1 sleeping for 0x3e8 ticks
Scheduler: Woke thread 0x1 0x30d3f cycles early
Timer : thread 0x1 Woke 1
Scheduler: Thread 0x1 sleeping for 0x3e8 ticks
Scheduler: Woke thread 0x1 0x30d3e cycles early
Timer : thread 0x1 Woke 2
```

So digging in the code I saw there's now a flags argument to control this, but even with that set
```
void __cheri_compartment("timer") init()
{
	int loop = 1;
	while (true)
	{
		Timeout t1{1000};
		thread_sleep(&t1, ThreadSleepNoEarlyWake);
		Debug::log("thread {} Woke {}", thread_id_get(), loop++);
	}
}
```
although the log no longer shows that the thread is waking early
```
Scheduler: Created thread for trusted stack 0x80000230 (v:1 0x80000230-0x80000338 l:0x108 o:0xa p: G RWcgml -- ---)
Scheduler: Thread 0x1 sleeping for 0x3e8 ticks
Timer : thread 0x1 Woke 1
Scheduler: Thread 0x1 sleeping for 0x3e8 ticks
Timer : thread 0x1 Woke 2
``` 
it still returns more or less immediately - I was expecting that telling it not to wake early would get me back to the previous behaviour.

Clearly I'm missing something , can someone point me  in the right direction please ?

BTW I did try using
``
   Timeout t1{MS_TO_TICKS(1000)};
``
But that seems to result in an even smaller value being passed into the sleep.

**Comment by davidchisnall** [comment]
This looks like a bug.  I would expect your code to sleep for at least 1000 ticks (which, on Sonata, is 10s, so `MS_TO_TICKS(1000)` would give 100 ticks).  Is this Sonata or some other target?

**Reply by davidchisnall** [reply]
We calculate the time for expiring timers [here](https://github.com/microsoft/cheriot-rtos/blob/main/sdk/core/scheduler/timer.h#L173C27-L173C57) and then expire them in the function above.  I wonder if we're seeing 32-bit integer overflow there and ending up with a very small time distance?  I don't think that should happen, but that area is probably near the bug.  I can try to reproduce this tomorrow morning and see.

**Reply by PhilDay-CT** [reply]
I'm running the dev container, so the target is Sail.

I'll take a look at that code and see if I can spot anything.

**Reply by davidchisnall** [reply]
Oh, okay.  In simulation, we elapse time very fast because time is not real.

**Reply by davidchisnall** [reply]
I hadn't considered demos in sail (there's also the Ibex sim there, but I think it has the same issue - we have some `#ifdef SIMULATION` things that just assume time is an illusion, lunchtime doubly so).

**Reply by davidchisnall** [reply]
FWIW, doing this on Sonata does sleep for one second between each tick.  

I'm a bit surprised it looked like it worked on Sail previously.  The Sail config pretends that the machine runs at 2000 instructions per second, with 10 ticks per second (so 200 instructions per tick).  A 1000 tick sleep is therefore 200,000 instructions, which is probably around half a second on my machine, so vaguely close but only by coincidence.

**Comment by PhilDay-CT** [comment]
It looks like timer.h has evolved quite a lot from the version I was previously working against.    Looking at the code it looks like the reason this is jumping directly to the next timer is because threadHasNoPeers is evaluating to true, 

https://github.com/microsoft/cheriot-rtos/blob/7f4c6101c7d7467ef12de37fa2d5ac2cb91cbfce/sdk/core/scheduler/timer.h#L80C3-L103C4

which means that nextTick gets set to DistantFuture and the clock seems to then jump directly to the value for the timer rather that going though the ticks inbetween ? 

If I force threadHasNoPeers to false then I get the timers working in realtime (rather than lunchtime).

**Reply by davidchisnall** [reply]
That happens only for yielding (without the flag). There’s a longer explanation of the changes here:

https://cheriot.org/scheduler/2024/06/07/tickless-scheduler.html

**Reply by davidchisnall** [reply]
Now I’m confused about why that would make a difference. We schedule a timer interrupt in two cases:

 - If there is a thread with equal priority to the current one, we schedule a timer so that we can round-robin schedule equal-priority threads.
 - If there is a thread blocked on a timeout, we schedule a timer for the next thread to wake (we could skip this if the thread is lower priority than the current one).

If both of these are true, we schedule the timer for whichever is sooner.

In your case, there is one thread (which is not currently running, so no peers), but there is a thread blocked on the timer, so we should be scheduling a timer interrupt for that timeout and then sleeping (scheduling the idle thread, which just does WFI) until that timer fires, then we should wake that thread. And, indeed, that is what happens on Sonata.

I don’t understand how setting no-peers to false would change this. If the timer is for less than one tick, it will do nothing, if it’s more than one tick then we’d get the timer interrupt sooner but not wake the thread because no timeout had expired.

**Reply by davidchisnall** [reply]
Can you add some logs in expiretimers and expiry_time_for_timeout to see what the sleep timeout is and what the time at wake is?

**Reply by PhilDay-CT** [reply]
Where I was hacking no-peers makes the call setnext(nextTick) instead of setnext(nextEvent).   With setnext(nextTick) the scheduler still behaves as the old tick based scheduler I guess,  so maybe the delays I'm seeing just come from the scheduler overhead of processing all the ticks where there is nothing to do ?   

With that extra tracing, and threadHasNoPeers hacked to be false:
```
Scheduler: Thread 0x1 sleeping for 0x3e8 ticks         <- Call to thread_sleep
Scheduler: Set timeout to 0x3e8
Scheduler: Returning 0x31224                           <- return value from expiry_time_for_timeout
Scheduler: now 0x4ed iter exp: 0x3122b                 <- expiretimers()
Scheduler: Next tick: :0x5c1 nextTimer: 0x3122b        <- nextTick and NextTimer values in Update() - NextTick passed to setnext()
Scheduler: now 0x507 iter exp: 0x3122b
Scheduler: Next tick: :0x5db nextTimer: 0x3122b
Scheduler: now 0x5dc iter exp: 0x3122b
Scheduler: Next tick: :0x6b1 nextTimer: 0x3122b
Scheduler: now 0x6b2 iter exp: 0x3122b
Scheduler: Next tick: :0x787 nextTimer: 0x3122
```

Without the Hack
```
Scheduler: Thread 0x1 sleeping for 0x3e8 ticks                   <- Call to thread_sleep
Scheduler: Set timeout to 0x3e8
Scheduler: Returning 0x31224                                     <- return value from expiry_time_for_timeout
Scheduler: now 0x4ec iter exp:                                   <- expiretimers()
Scheduler: Next tick: :0xffffffffffffffff nextTimer: 0x3122b     <- nextTick and NextTimer values in Update() - NextTimer passed to setnext()
Scheduler: now 0x509 iter exp: 0x3122b                           <- expiretimers()     
Scheduler: Next tick: :0xffffffffffffffff nextTimer: 0x3122b
Scheduler: now 0x3122c iter exp: 0x3122b                         <- clock "jumps" directly to 0x3122c
Scheduler: Wake up !!!
Timer : thread 0x1 Woke 1

``` 

Its almost as if on Sail setnext() just jumps the clock directly to that value  - what is it using for the underlying timer calculation if there are no instructions to execute ?

**Reply by rmn30** [reply]
In sail the `wfi` instruction sets `mtime` to the timer value so that the timer immediately triggers (Sail has no other interrupt sources). I note that this might break the scheduler's assumption that a 64-bit timer will never overflow although I don't think that's what's happening here.

**Reply by davidchisnall** [reply]
> maybe the delays I'm seeing just come from the scheduler overhead of processing all the ticks where there is nothing to do ?

That's possible.  The 2000-instruction tick time on Sail is very short.  The scheduler is probably using a very large fraction of that.

> Its almost as if on Sail setnext() just jumps the clock directly to that value - what is it using for the underlying timer calculation if there are no instructions to execute ?

Ah, yes, I forgot about that.  When you do WFI on Sail, it skips immediately to the next timer interrupt.  It's an architectural model with no notion of real time (or a pipeline to simulate cycles elapsing) and so, when there are no instructions to retire it just advances to the next point where there are.

I believe the ibex sim (also in the dev container) will do what you expect, because it does have a pipeline model and so will simulate a pipeline doing nothing for the required number of cycles.

**Reply by PhilDay-CT** [reply]
So no way on sail to achieve an idle wait now then with the tickless scheduler ?

**Reply by davidchisnall** [reply]
There never was, you just had an accidental busy-wait loop that looked a bit like you had one (that would go away if the Sail JSON had more sensible values for the tick length - 2000 instructions is very silly).  You can do a busy wait, or you can use the Ibex simulator if you want something that looks a bit more like a real system.  Sail has no input sources or notion of time, Ibex sim has both (the UART works for reading and writing) and so is a better choice for demos.

**Reply by rmn30** [reply]
> So no way on sail to achieve an idle wait now then with the tickless scheduler ?

I don't think so, sorry. Could you use a busy wait instead? Seems like it only worked previously because the ticky scheduler was so inefficient!

If you really want we could potentially add an option to disable the Sail `wfi` optimisation, but probably better to run on Ibex.

**Reply by PhilDay-CT** [reply, marked as answer]
@rmn30 - Running on ibex if fine in this case ;-)

**Reply by rmn30** [reply]
> that would go away if the Sail JSON had more sensible values for the tick length - 2000 instructions is very silly

As a side note this is not quite correct. Sail time also has another quirk, which is that `mcycle` and `mtime` actually increment every 100 instructions (Sail has IPC of 100!). I think this means the current `timer_hz` of 2000 and `tickrate_hz` of 10 in `sail.json` gives about 20000 instructions per timer tick. This is probably still a bit low.

**Comment by PhilDay-CT** [comment]
Yep - ibex sim works fine ;-)

Thanks for all the patience and explanations
